### PR TITLE
Fix parser cache files maintenance

### LINKED ISSUE #7061: Suboptimal maintenance of parser cache files using `touch()`
The modification time of files in the parser cache directory (`ParserCacheProvider::PARSER_CACHE_DIRECTORY`, `php-parser`) may be changed multiple times during a single run:

- in `ProjectAnalyzer::check()` (on [full](https://github.com/vimeo/psalm/blob/14dcbc908ab2625cd7a74258ee6c740cbecc6140/src/Psalm/Internal/Cli/Psalm.php#L358), [diff](https://github.com/vimeo/psalm/blob/14dcbc908ab2625cd7a74258ee6c740cbecc6140/src/Psalm/Internal/Analyzer/ProjectAnalyzer.php#L680) run):
  https://github.com/vimeo/psalm/blob/14dcbc908ab2625cd7a74258ee6c740cbecc6140/src/Psalm/Internal/Analyzer/ProjectAnalyzer.php#L681

  by calling [`ParserCacheProvider::touchParserCaches()`](https://github.com/vimeo/psalm/blob/14dcbc908ab2625cd7a74258ee6c740cbecc6140/src/Psalm/Internal/Provider/ParserCacheProvider.php#L378) with the results of [`ProjectAnalyzer::getAllFiles()`](https://github.com/vimeo/psalm/blob/14dcbc908ab2625cd7a74258ee6c740cbecc6140/src/Psalm/Internal/Analyzer/ProjectAnalyzer.php#L1062), which may be particularly expensive, as the latter iterates recursively over all project directories, including unrelated files (see also: #7057)

- in `IssueBuffer::finish()` (on [full](https://github.com/vimeo/psalm/blob/14dcbc908ab2625cd7a74258ee6c740cbecc6140/src/Psalm/IssueBuffer.php#L724) run):
https://github.com/vimeo/psalm/blob/14dcbc908ab2625cd7a74258ee6c740cbecc6140/src/Psalm/IssueBuffer.php#L732

  by calling [`ParserCacheProvider::processSuccessfulRun()`](https://github.com/vimeo/psalm/blob/14dcbc908ab2625cd7a74258ee6c740cbecc6140/src/Psalm/Internal/Provider/ParserCacheProvider.php#L350) which uses a simple loop over `scandir($cache_directory)`

Given that the only non-maintenance usage of modification time appears to be the comparison with source files at:
https://github.com/vimeo/psalm/blob/14dcbc908ab2625cd7a74258ee6c740cbecc6140/src/Psalm/Internal/Provider/ParserCacheProvider.php#L100

the apparent reason is to allow purging of old cache files in `ProjectAnalyzer::check()` at:
https://github.com/vimeo/psalm/blob/14dcbc908ab2625cd7a74258ee6c740cbecc6140/src/Psalm/Internal/Analyzer/ProjectAnalyzer.php#L672-L674
by calling [`ParserCacheProvider::deleteOldParserCaches()`](https://github.com/vimeo/psalm/blob/14dcbc908ab2625cd7a74258ee6c740cbecc6140/src/Psalm/Internal/Provider/ParserCacheProvider.php#L318), which seems to be non-functional:
https://github.com/vimeo/psalm/blob/b359847b749658fe385bdea662331146845c0724/src/Psalm/Internal/Provider/ParserCacheProvider.php#L322-L324

The mechanism that relies on `touch()` (likely introduced via https://github.com/vimeo/psalm/commit/e9b4eb69e5d93da645d6b6ac5a931030aeffa919) could be replaced completely with one that deletes all cache files before a full run, or one that relies on a list of processed files for which a cache should be kept.

### PR DESCRIPTION
**Pull request by dvz**
Resolves #7061

### CODE REVIEW

File: src/Psalm/Internal/Analyzer/ProjectAnalyzer.php [hunk 1]
```diff
@@ -1074,6 +1068,7 @@ private function checkDirWithConfig(string $dir_name, Config $config, bool $allo
 
     /**
      * @return list<string>
+     * @deprecated going to be removed in Psalm 5
```
Comment (weirdan) [inline src/Psalm/Internal/Analyzer/ProjectAnalyzer.php:1071]: Private function of an internal class may be removed right away.

File: src/Psalm/Internal/Provider/ParserCacheProvider.php [hunk 2]
```diff
@@ -349,6 +349,9 @@ public function deleteOldParserCaches(float $time_before): int
         return $removed_count;
     }
 
+    /**
+     * @deprecated going to be removed in Psalm 5
+     */
     public function processSuccessfulRun(): void
```
Comment (weirdan) [inline src/Psalm/Internal/Provider/ParserCacheProvider.php:355]: This doesn't require deprecation period either.

File: src/Psalm/Internal/Provider/ParserCacheProvider.php [hunk 3]
```diff
@@ -376,6 +379,7 @@ public function processSuccessfulRun(): void
 
     /**
      * @param  array<string>    $file_names
+     * @deprecated going to be removed in Psalm 5
      */
     public function touchParserCaches(array $file_names, int $min_time): void
```
Comment (weirdan) [inline src/Psalm/Internal/Provider/ParserCacheProvider.php:384]: ditto

**Comment by orklah** [comment]
Thanks!

**Pull request merged**